QuestionAnswering/QA_ICL/README.md

```diff
@@ -1,15 +1,130 @@
-# Question-Answering for The World Avatar using In-Context Learning
+# Question-Answering System for The World Avatar
+
+## Table of Contents
+- [Question-Answering System for The World Avatar](#question-answering-system-for-the-world-avatar)
+  - [Table of Contents](#table-of-contents)
+  - [Introduction](#introduction)
+    - [Key Features](#key-features)
+  - [Architecture overview](#architecture-overview)
+  - [Key Components](#key-components)
+  - [Project Structure](#project-structure)
+  - [Getting Started](#getting-started)
+    - [Prerequisities](#prerequisities)
+    - [Installation](#installation)
+    - [API Documentation](#api-documentation)
+    - [Usage](#usage)
 
 ## Introduction
 
-Question-answering for The World Avatar involves retrieving data from RDF graphs and other data sources such as HTTP endpoints. To do so, input questions need to be converted into SPARQL queries and HTTP requests, whose execution would yield the desired data. The conversion of natural language queries to data requests is facilitated by in-context learning, which entails engineering a text prompt for LLMs to automatically perform the transformation. The prompt may include context information such as parsing examples and the structure of target predictions.
+The Question-Answering System for The World Avatar involves retrieving data from RDF graphs and other data sources such as HTTP endpoints. To do so, input questions need to be converted into SPARQL queries and HTTP requests, whose execution would yield the desired data. The conversion of natural language queries to data requests is facilitated by in-context learning (ICL), which entails engineering a text prompt for Large Language Models (LLMs) to automatically perform the transformation. The prompt may include context information such as parsing examples and the structure of target predictions.
+
+### Key Features
+- Utilises in-context learning for query transformation
+- Supports multiple data sources including RDF graphs and HTTP endpoints
+- Employs LLMs for natural language understanding
+- Provides a user-friendly frontend interface
+- Offers efficient backend processing with caching and inference capabilities
+- Includes tools for data generation and preparation
 
 ## Architecture overview
 
-![](uml/system_architecture.png)
+The architecture of the project consists of a frontend (user interface), a backend (handling queries, model inference, and Knowledge Graph access), and additional data generation modules. These components communicate through RESTful APIs and other protocols. 
+
+```mermaid
+graph TD
+    subgraph Frontend
+    A[Next.js app]
+    end
+
+    subgraph Backend
+    C[FastAPI app]
+    E[(Redis)]
+    F[Triton inference server]
+    end
+
+    D[(Blazegraph servers)]
+
+    A <--> |HTTP| C
+    D <--> |SPARQL protocol| C
+    C <--> |REST| E
+    C <--> |gRPC| F
```

Comment:
What I originally intended for this diagram is to show the different components that make up Marie and the interface between them. With this latest diagram, it is confusing because it is trying to depict the events that take place between the components, but it doesn't show the ordered nature of this sequence of events.
My previous suggestion for you when sticking to Mermaid is to have a single unidirectional arrow between any two components, with the arrow label being the interface for the communication (HTTP, gRPC, RESP, SPARQL). Then, you have an annotation for the meaning of the arrow direction e.g. direction of interface consumption or provision. Optionally, you can include a short description of the role of each component in the component box itself e.g. "Redis (search and expirable storage)"